api/src/oscc.c

```diff
@@ -492,110 +603,372 @@ oscc_result_t oscc_async_enable( int socket )
+}
+
+
+int init_can_socket( const char *can_channel,
+                     struct timeval *tv,
+                     struct can_filter *filter )
+{
+    int valid = -1;
+    int sock = -1;
     struct ifreq ifr;
-    memset( &ifr, 0, sizeof(ifr) );
+    memset( &ifr, 0, sizeof( ifr ) );
 
-    if ( result == OSCC_OK )
+    sock = socket( PF_CAN, SOCK_RAW, CAN_RAW );
+
+    if ( sock < 0 )
+    {
+        perror( "Opening CAN socket failed:" );
+    }
+    else
     {
         strncpy( ifr.ifr_name, can_channel, IFNAMSIZ );
 
-        ret = ioctl( sock, SIOCGIFINDEX, &ifr );
+        valid = ioctl( sock, SIOCGIFINDEX, &ifr );
 
-        if ( ret < 0 )
+        if ( valid < 0 )
         {
-            printf( "Finding CAN index failed: %s\n", strerror(errno) );
+            perror( "Finding CAN index failed:" );
+        }
+    }
 
-            result = OSCC_ERROR;
+    //If a timeout has been specified set one here since it should be set before
+    //the bind call
+    if( valid >= 0 && tv != NULL )
+    {
+        valid = setsockopt( sock,
+                            SOL_SOCKET,
+                            SO_RCVTIMEO,
+                            tv,
+                            sizeof( struct timeval ) );
+
+        if ( valid < 0 )
+        {
+            perror( "Setting timeout failed:" );
         }
     }
 
+    //If a filter has been specified set one here
+    if( (valid >= 0) && (filter != NULL) )
+    {
+        int ret = setsockopt( sock,
+                    SOL_CAN_RAW,
+                    CAN_RAW_FILTER,
+                    &filter,
+                    sizeof( filter ) );
 
-    if ( result == OSCC_OK )
+        if ( valid < 0 )
+        {
+            perror( "Setting filter failed:" );
+        }
+    }
+
+    if ( valid >= 0 )
     {
         struct sockaddr_can can_address;
 
-        memset( &can_address, 0, sizeof(can_address) );
+        memset( &can_address, 0, sizeof( can_address ) );
         can_address.can_family = AF_CAN;
         can_address.can_ifindex = ifr.ifr_ifindex;
 
-        ret = bind(
-            sock,
-            (struct sockaddr *) &can_address,
-            sizeof(can_address) );
+        valid = bind( sock,
+                      (struct sockaddr *) &can_address,
+                      sizeof( can_address ) );
 
-        if ( ret < 0 )
+        if ( valid < 0 )
         {
-            printf( "Socket binding failed: %s\n", strerror(errno) );
-
-            result = OSCC_ERROR;
+            perror( "Socket binding failed:" );
         }
     }
 
+    // Clean up resources and close the connection if it's invalid.
+    if( valid < 0 )
+    {
+      close( sock );
+      sock = -1;
+    }
 
-    if ( result == OSCC_OK )
+    return sock;
+}
+
+
+struct can_contains can_detection( const char *can_channel )
+{
+    struct timeval timeout;
+    timeout.tv_sec = 0;
+    timeout.tv_usec = CAN_MESSAGE_TIMEOUT;
+
+    int sock = init_can_socket( can_channel, &timeout, NULL );
+
+    int i = 0;
```

Comment:
Nitpick: Can we use the C99 feature to declare variables inside of a for loop initializing expression? If not, can we move this variable declaration closer to the `for` loop that uses it?

Reply:
Let's park the C99 discussion for a larger scope discussion. I'll move it closer for now.